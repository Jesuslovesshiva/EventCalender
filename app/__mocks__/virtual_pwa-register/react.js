import {useState} from 'react'

const useRegisterSW = jest.fn(() => {
  const [needRefresh, setNeedRefresh] = useState(true)

  return {
    needRefresh: [needRefresh, setNeedRefresh],
    updateServiceWorker: () => {}
  }
})

// eslint-disable-next-line import/prefer-default-export
export {useRegisterSW}
