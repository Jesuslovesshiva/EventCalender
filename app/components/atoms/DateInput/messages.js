import {defineMessages} from 'react-intl'

const scope = 'atoms.date-picker'

export default defineMessages({
  ddmmyyyy: {
    id: `${scope}.date`,
    defaultMessage: 'TT.MM.JJJJ',
    description: 'placeholder for format dd.MM.yyyy'
  }
})
