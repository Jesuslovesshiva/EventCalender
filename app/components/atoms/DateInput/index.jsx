import {format, isValid, parse} from 'date-fns'
import noop from 'lodash/noop'
import PropTypes from 'prop-types'
import {forwardRef, useCallback, useEffect, useState} from 'react'
import {Rifm} from 'rifm'

import useTranslation from '@/hooks/use-translation'

import messages from './messages'

const FORMAT = 'dd.MM.yyyy'
const FORMAT_MESSAGE_KEY = FORMAT.toLowerCase()
  .replaceAll(/[^a-z]/gu, '')
  .toLowerCase()
const DELIMITER_INDICES = [...FORMAT.matchAll(/\./gu)].map(match => match.index)

const maskAppend = input => {
  if (DELIMITER_INDICES.includes(input.length)) {
    return `${input}.`
  }

  return input
}

const maskReplace = input => input.replaceAll('Z', '0')

const maskFormatPrependZeros = (input, wip) => {
  let output = wip

  if (output.length === DELIMITER_INDICES[0] - 1) {
    output = `Z${output}`
  }

  if (
    output.length === DELIMITER_INDICES[1] - 1 &&
    output[DELIMITER_INDICES[0]] === '.'
  ) {
    // Get sliceIndex by removing one from index for the delimiter and one for missing zero
    // eslint-disable-next-line no-magic-numbers
    const sliceIndex = DELIMITER_INDICES[1] - 2
    output = `${output.slice(0, sliceIndex)}Z${output.slice(sliceIndex)}`
  }

  return output
}

const maskFormat = input => {
  const digits = input.match(/\d/gu) ?? []
  let output = digits
    .reduce((acc, digit, index) => {
      if (
        index === DELIMITER_INDICES[0] ||
        index === DELIMITER_INDICES[1] - 1
      ) {
        return `${acc}.${digit}`
      }
      return `${acc}${digit}`
    }, '')
    .substr(0, FORMAT.length)

  if (input.endsWith('.')) {
    output = maskFormatPrependZeros(input, output)

    // istanbul ignore else
    if (DELIMITER_INDICES.includes(output.length)) {
      output += '.'
    }
  }

  return output
}

const DateInput = forwardRef((props, ref) => {
  const [inputValue, setInputValue] = useState('')
  const placeholder = useTranslation(messages[FORMAT_MESSAGE_KEY] ?? '-')

  const resetInputValue = useCallback(() => {
    try {
      setInputValue(format(props.value, FORMAT))
    } catch (error) {
      // Noop
    }
  }, [setInputValue, props.value])

  const update = useCallback(
    value => {
      const newDate = parse(value, FORMAT, new Date())

      if (isValid(newDate)) {
        props.onChange(newDate)
      } else {
        resetInputValue()
      }
    },
    [props.onChange, resetInputValue]
  )

  const onBlur = useCallback(
    event => {
      update(event.target.value)
      props.onBlur()
    },
    [update, props.onBlur]
  )

  const onKeyDown = useCallback(
    event => {
      if (event.key === 'Enter') {
        update(event.target.value)
        event.target.blur()
        props.onSubmit()
      }
    },
    [update, props.onSubmit]
  )

  useEffect(() => {
    resetInputValue()
  }, [props.value])

  return (
    <Rifm
      mask={inputValue.length >= FORMAT.length}
      value={inputValue}
      onChange={setInputValue}
      format={maskFormat}
      replace={maskReplace}
      append={maskAppend}
    >
      {({onChange, value}) => (
        <input
          ref={ref}
          placeholder={placeholder}
          value={value}
          onChange={onChange}
          aria-label={props['aria-label']}
          onFocus={props.onFocus}
          onBlur={onBlur}
          onKeyDown={onKeyDown}
          className="
            w-[75px] rounded-sm transition-[outline-color,background-color]
            bg-transparent font-bold text-center
            outline-transparent focus-visible:bg-silver focus-visible:outline
            focus-visible:outline-1 focus-visible:outline-gray-medium focus-visible:outline-offset-0
          "
        />
      )}
    </Rifm>
  )
})

DateInput.displayName = 'DateInput'

DateInput.propTypes = {
  'aria-label': PropTypes.string,
  onBlur: PropTypes.func,
  onChange: PropTypes.func,
  onFocus: PropTypes.func,
  onSubmit: PropTypes.func,
  value: PropTypes.instanceOf(Date)
}

DateInput.defaultProps = {
  'aria-label': null,
  onBlur: noop,
  onChange: noop,
  onFocus: noop,
  onSubmit: noop,
  value: null
}

export default DateInput
