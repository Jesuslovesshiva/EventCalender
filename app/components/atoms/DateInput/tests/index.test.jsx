import {fireEvent} from '@testing-library/react'

import render from '@/utils/test-utils/custom-render'

import DateInput from '../index'

const renderComponent = (props, renderFn = render) =>
  renderFn(<DateInput {...props} />)

describe('DateCyclePicker', () => {
  describe('Input', () => {
    it('should match the snapshot', () => {
      const {container} = renderComponent({value: new Date(1)})
      expect(container.firstChild).toMatchSnapshot()
    })

    it('should update input value on value prop change', () => {
      const {container, rerender} = renderComponent({value: new Date(1)})
      expect(container.firstChild.value).toBe('01.01.1970')
      renderComponent({value: new Date(1646735979483)}, rerender)
      expect(container.firstChild.value).toBe('08.03.2022')
    })

    describe('mask', () => {
      it('should append a . after day input', () => {
        const {container} = renderComponent()
        fireEvent.change(container.firstChild, {target: {value: '28'}})
        expect(container.firstChild.value).toBe('28.')
      })

      it('should append a . after month input', () => {
        const {container} = renderComponent()
        fireEvent.change(container.firstChild, {target: {value: '28.03'}})
        expect(container.firstChild.value).toBe('28.03.')
      })

      it('should add a leading zero to 1-digit day', () => {
        const {container} = renderComponent()
        fireEvent.change(container.firstChild, {target: {value: '1.'}})
        expect(container.firstChild.value).toBe('01.')
      })

      it('should add a leading zero to 1-digit month', () => {
        const {container} = renderComponent()
        fireEvent.change(container.firstChild, {target: {value: '01.1.'}})
        expect(container.firstChild.value).toBe('01.01.')
      })
    })

    describe('events', () => {
      it('should call onChange on blur', () => {
        const onChange = jest.fn()
        const {container} = renderComponent({onChange, value: new Date(1)})

        expect(onChange).not.toHaveBeenCalled()
        fireEvent.change(container.firstChild, {target: {value: '08.03.2022'}})
        fireEvent.blur(container.firstChild)
        expect(onChange).toHaveBeenCalledWith(new Date(2022, 2, 8))
      })
      it('should call onChange and blur on Enter key down', () => {
        const onChange = jest.fn()
        const {container} = renderComponent({onChange, value: new Date(1)})
        container.firstChild.blur = jest.fn()
        expect(container.firstChild.blur).not.toHaveBeenCalled()
        expect(onChange).not.toHaveBeenCalled()
        fireEvent.keyDown(container.firstChild, {key: 'Enter'})
        expect(container.firstChild.blur).toHaveBeenCalled()
        expect(onChange).toHaveBeenCalled()
      })

      it('should not call onChange and blur on any other key down', () => {
        const onChange = jest.fn()
        const {container} = renderComponent({onChange, value: new Date(1)})
        container.firstChild.blur = jest.fn()
        expect(container.firstChild.blur).not.toHaveBeenCalled()
        expect(onChange).not.toHaveBeenCalled()
        fireEvent.keyDown(container.firstChild, {key: 'ArrowRight'})
        expect(container.firstChild.blur).not.toHaveBeenCalled()
        expect(onChange).not.toHaveBeenCalled()
      })

      it('should reset input value on confirm of invalid date input', () => {
        const onChange = jest.fn()
        const {container} = renderComponent({onChange, value: new Date(1)})

        expect(onChange).not.toHaveBeenCalled()
        fireEvent.change(container.firstChild, {target: {value: 'foo'}})
        fireEvent.blur(container.firstChild)
        expect(onChange).not.toHaveBeenCalled()
        expect(container.firstChild.value).toBe('01.01.1970')
      })
    })
  })
})
