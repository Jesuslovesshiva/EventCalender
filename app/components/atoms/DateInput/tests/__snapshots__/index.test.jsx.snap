// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateCyclePicker Input should match the snapshot 1`] = `
<input
  class=" w-[75px] rounded-sm transition-[outline-color,background-color] bg-transparent font-bold text-center outline-transparent focus-visible:bg-silver focus-visible:outline focus-visible:outline-1 focus-visible:outline-gray-medium focus-visible:outline-offset-0 "
  placeholder="TT.MM.JJJJ"
  value="01.01.1970"
/>
`;
